core/src/main/java/org/bf2/srs/fleetmanager/execution/impl/workers/provision/CreateSubscriptionWorker.java

```diff
@@ -0,0 +1,174 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.inject.Inject;
+import javax.transaction.Transactional;
+import java.util.List;
+import java.util.Optional;
+
+import static org.bf2.srs.fleetmanager.execution.impl.tasks.TaskType.CREATE_SUBSCRIPTION_T;
+import static org.bf2.srs.fleetmanager.execution.impl.workers.WorkerType.CREATE_SUBSCRIPTION_W;
+
+/**
+ * This class MUST be thread safe. It should not contain state and inject thread safe beans only.
+ */
+@ApplicationScoped
+public class CreateSubscriptionWorker extends AbstractWorker {
+
+    private final Logger log = LoggerFactory.getLogger(getClass());
+
+    @Inject
+    ResourceStorage storage;
+
+    @Inject
+    AccountManagementService accountManagementService;
+
+    @ConfigProperty(name = "srs-fleet-manager.registry.instances.eval.enabled")
+    boolean evalInstancesEnabled;
+
+    @ConfigProperty(name = "srs-fleet-manager.registry.instances.eval.only")
+    boolean evalInstancesOnlyEnabled;
+
+    @ConfigProperty(name = "srs-fleet-manager.registry.instances.eval.max-count-per-user")
+    int maxEvalInstancesPerUser;
+
+    @Inject
+    TaskManager tasks;
+
+    public CreateSubscriptionWorker() {
+        super(CREATE_SUBSCRIPTION_W);
+    }
+
+    @Override
+    public boolean supports(Task task) {
+        return CREATE_SUBSCRIPTION_T.name().equals(task.getType());
+    }
+
+    @Override
+    @Transactional
+    public void execute(Task aTask, WorkerContext ctl) throws Exception {
+        CreateSubscriptionTask task = (CreateSubscriptionTask) aTask;
+        Optional<RegistryData> registryOptional = storage.getRegistryById(task.getRegistryId());
+
+        if (registryOptional.isEmpty()) {
+            // NOTE: Failure point 1
+            ctl.retry();
+        }
+
+        AccountInfo accountInfo = task.getAccountInfo();
+        RegistryData registry = registryOptional.get();
+
+        //Only try create a subscription if there is account information attached to the task.
+        boolean failedProvisioning = false;
+        String provisioningError = null;
+        if (task.getAccountInfo() != null) {
+
+            // Figure out if we are going to create a standard or eval instance.
+            ResourceType resourceType = determineResourceType(accountInfo);
+
+            // Try to consume some quota from AMS for the appropriate resource type (standard or eval).  If successful
+            // we'll get back a subscriptionId - if not we'll throw an exception.
+
+            try {
+                String subscriptionId = accountManagementService.createResource(accountInfo, resourceType);
+                registry.setSubscriptionId(subscriptionId);
+            } catch (ResourceLimitReachedException rlre) {
+                log.warn("Resource limit reached:", rlre);
+                failedProvisioning = true;
+                provisioningError = rlre.getMessage();
```

Comment:
For user messages, this should be used `rlre.getUserErrorInfo().getReason()`. I would prefer to include the whole `Error` entity in that field, which would enable including `rlre.getUserErrorInfo().getCode()` and `operationId` as well. The entity could be serialized into a string in the DB.